File: Backend/src/main/java/com/Bookie/entities/ChapterEntity.java
package com.Bookie.entities;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;

@Entity(name = "ChapterEntity")
@Table(name = "chapter")
@Getter
@Setter
@Builder
@AllArgsConstructor
@NoArgsConstructor
@EqualsAndHashCode(of = "id")
public class ChapterEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT", nullable = false)
    private String content;

    @ManyToOne
    @JoinColumn(name = "history_id", nullable = false)
    @JsonIgnore
    private HistoryEntity history;

    @Column(nullable = false)
    private Double latitude;

    @Column(nullable = false)
    private Double longitude;

    @Column(name = "img", columnDefinition = "text", length = 1000)
    private String img;
}

File: frontend/lib/domain/entities/chapter_entity.dart
class ChapterPartial {
  final int id;
  final String title;
  final double latitude;
  final double longitude;
  String? image;

  ChapterPartial({
    required this.id,
    required this.title,
    required this.latitude,
    required this.longitude,
    this.image,
  });

  Map<String, dynamic> toJson() {
    return {
      "id": id,
      "title": title,
      "latitude": latitude,
      "longitude": longitude,
      "image": image
    };
  }
}

class Chapter extends ChapterPartial {
  final String content;
  int? historyId;

  Chapter({
    required this.content,
    required super.id,
    required super.title,
    required super.latitude,
    required super.longitude,
    this.historyId,
    super.image,
  });

  @override
  String toString() {
    return 'Chapter(id: $id, title: $title, content: $content, historyId: $historyId, latitude: $latitude, longitude: $longitude, image: $image)';
  }
}

class ChapterForm {
  final int historyId;
  final String content;
  final String title;
  final double latitude;
  final double longitude;
  String? image;

  ChapterForm({
    required this.historyId,
    required this.title,
    required this.latitude,
    required this.longitude,
    required this.content,
    this.image,
  });

  Map<String, dynamic> toJson() {
    return {
      "historyId": historyId,
      "title": title,
      "latitude": latitude,
      "longitude": longitude,
      "content": content,
      "image": image
    };
  }
}
